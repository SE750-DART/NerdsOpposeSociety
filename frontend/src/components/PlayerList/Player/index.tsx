import React from "react";
import styles from "./style.module.css";

interface PlayerProps {
  nickname: string;
  score?: number;
  highlight?: boolean;
  divider: boolean;
  isHost?: boolean;
}

const Player = ({
  nickname,
  score,
  highlight = false,
  divider,
  isHost,
}: PlayerProps) => {
  let highlightStyle = "";
  if (highlight) {
    highlightStyle = styles.playerMe;
  }

  return (
    <div className={`${styles.player} ${highlightStyle}`}>
      <div className={styles.playerText}>{nickname}</div>
      {isHost && <div className={styles.playerIndicator}>HOST</div>}
      <div className={styles.playerScore}>{score}</div>
      {divider && <div className={styles.playerDivider} />}
    </div>
  );
};

export default Player;
