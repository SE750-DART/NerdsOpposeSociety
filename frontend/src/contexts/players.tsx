import React, { useState } from "react";
import useCrud from "../hooks/useCrud";

export type Player = {
  nickname: string;
  id: string;
  score: number;
};
const equals = (player1: Player, player2: Player) => player1.id === player2.id;

type Context = {
  host: string;
  setHost: (host: string) => void;
  players: Player[];
  initialisePlayers: (players: Player[]) => void;
  addPlayer: (id: string, nickname: string) => void;
  removePlayer: (playerId: string) => void;
  incrementPlayerScore: (playerId: string) => void;
};

const PlayersContext = React.createContext<Context>({
  host: "",
  setHost: () => null,
  players: [],
  initialisePlayers: () => null,
  addPlayer: () => null,
  removePlayer: () => null,
  incrementPlayerScore: () => null,
});

const PlayersContextProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [host, setHost] = useState("");

  const {
    items: players,
    initialiseItems: initialisePlayers,
    addItem,
    removeItem,
    updateItem,
  } = useCrud<Player>(equals);

  const addPlayer = (id: string, nickname: string) =>
    addItem({
      nickname,
      id,
      score: 0,
    });

  const removePlayer = (playerId: string) => {
    const playerToRemove = players.find((player) => player.id === playerId);
    if (!playerToRemove) {
      return;
    }

    removeItem(playerToRemove);
  };

  const incrementPlayerScore = (playerId: string) => {
    const playerToUpdate = players.find((player) => player.id === playerId);
    if (!playerToUpdate) {
      return;
    }

    updateItem({
      ...playerToUpdate,
      score: playerToUpdate.score + 1,
    });
  };

  // The context value that will be supplied to any descendants of this component.
  const context = {
    host,
    setHost: (newHost: string) => setHost(newHost),
    players,
    initialisePlayers,
    addPlayer,
    removePlayer,
    incrementPlayerScore,
  };

  // Wraps the given child components in a Provider for the above context.
  return (
    <PlayersContext.Provider value={context}>
      {children}
    </PlayersContext.Provider>
  );
};

export { PlayersContext, PlayersContextProvider };
