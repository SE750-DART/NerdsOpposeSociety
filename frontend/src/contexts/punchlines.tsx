import React from "react";
import useCrud from "../hooks/useCrud";

export type Punchline = {
  text: string;
  new?: boolean;
  viewed?: boolean;
};

const equals = (punchline1: Punchline, punchline2: Punchline) =>
  punchline1.text === punchline2.text;

type Context = {
  punchlines: Punchline[];
  initialisePunchlines: (arg0: Punchline[]) => void;
  addPunchlines: (arg0: string[]) => void;
  removePunchline: (arg0: string) => void;
};

const PunchlinesContext = React.createContext<Context>({
  punchlines: [],
  initialisePunchlines: () => null,
  addPunchlines: () => null,
  removePunchline: () => null,
});

const PunchlinesContextProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const {
    items: punchlines,
    initialiseItems: initialisePunchlines,
    addItems,
    removeItem,
    updateItem,
  } = useCrud<Punchline>(equals);

  const addPunchlines = (newPunchlines: string[]) => {
    if (punchlines.length === 0) {
      addItems(
        newPunchlines.map((punchline) => ({
          text: punchline,
          new: false,
          viewed: false,
        }))
      );
      return;
    }

    punchlines.forEach((punchline) => {
      updateItem({
        ...punchline,
        new: false,
        viewed: false,
      });
    });

    addItems(
      newPunchlines.map((punchline) => ({
        text: punchline,
        new: true,
        viewed: false,
      }))
    );
  };

  const removePunchline = (punchline: string) => {
    const punchlineToRemove = punchlines.find(
      (p: Punchline) => p.text === punchline
    );
    if (!punchlineToRemove) {
      return;
    }

    removeItem(punchlineToRemove);
  };

  const context = {
    punchlines,
    initialisePunchlines,
    addPunchlines,
    removePunchline,
  };

  return (
    <PunchlinesContext.Provider value={context}>
      {children}
    </PunchlinesContext.Provider>
  );
};

export { PunchlinesContext, PunchlinesContextProvider };
