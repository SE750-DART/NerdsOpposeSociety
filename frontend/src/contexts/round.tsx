import React, {
  Dispatch,
  PropsWithChildren,
  useContext,
  useReducer,
} from "react";
import { Punchline, Setup, Winner } from "../types";

export enum RoundAction {
  NEW_ROUND = "new_round",
  SET_SETUP = "set_setup",
  INCREMENT_PLAYERS_CHOSEN = "increment_players_chosen",
  SET_CHOSEN_PUNCHLINES = "set_chosen_punchlines",
  MARK_PUNCHLINE_READ = "mark_punchline_read",
  SET_WINNER = "set_winner",
}

type RoundActionType =
  | { type: RoundAction.NEW_ROUND; roundNumber: number }
  | { type: RoundAction.SET_SETUP; setup: Setup }
  // TODO: Combine NEW_ROUND & SET_SETUP
  | { type: RoundAction.INCREMENT_PLAYERS_CHOSEN }
  | { type: RoundAction.SET_CHOSEN_PUNCHLINES; punchlines: string[][] }
  | { type: RoundAction.MARK_PUNCHLINE_READ; index: number }
  | {
      type: RoundAction.SET_WINNER;
      winningPlayerId: string;
      winningPunchlines: string[];
    };

type RoundState = {
  roundNumber: number;
  setup: Setup | undefined;
  numPlayersChosen: number;
  chosenPunchlines: Punchline[];
  winner: Winner | undefined;
};

const reducer = (state: RoundState, action: RoundActionType): RoundState => {
  switch (action.type) {
    case RoundAction.NEW_ROUND: {
      const { roundNumber } = action;
      return {
        roundNumber,
        setup: undefined,
        numPlayersChosen: 0,
        chosenPunchlines: [],
        winner: undefined,
      };
    }

    case RoundAction.SET_SETUP: {
      const { setup } = action;
      return { ...state, setup };
    }

    case RoundAction.INCREMENT_PLAYERS_CHOSEN: {
      return { ...state, numPlayersChosen: state.numPlayersChosen + 1 };
    }

    case RoundAction.SET_CHOSEN_PUNCHLINES: {
      const { punchlines } = action;
      const chosenPunchlines = punchlines.map((playerPunchlines) => ({
        text: playerPunchlines[0],
        viewed: false,
      }));
      return { ...state, chosenPunchlines };
    }

    case RoundAction.MARK_PUNCHLINE_READ: {
      const { index } = action;
      const chosenPunchlines = [...state.chosenPunchlines];
      if (0 <= index && index < chosenPunchlines.length) {
        chosenPunchlines[index].viewed = true;
        return { ...state, chosenPunchlines };
      }
      return state;
    }

    case RoundAction.SET_WINNER: {
      const { winningPlayerId, winningPunchlines } = action;
      return { ...state, winner: { winningPlayerId, winningPunchlines } };
    }
  }
};

type RoundContextType = [RoundState, Dispatch<RoundActionType>];

const RoundContext = React.createContext<RoundContextType | undefined>(
  undefined
);

export const RoundProvider = ({ children }: PropsWithChildren<unknown>) => {
  const context = useReducer(reducer, {
    roundNumber: 0,
    setup: undefined,
    numPlayersChosen: 0,
    chosenPunchlines: [],
    winner: undefined,
  });

  return (
    <RoundContext.Provider value={context}>{children}</RoundContext.Provider>
  );
};

export const useRound = (): RoundContextType => {
  const context = useContext(RoundContext);

  if (context === undefined) {
    throw new Error("useRound() must be used within a RoundProvider");
  }

  return context;
};
