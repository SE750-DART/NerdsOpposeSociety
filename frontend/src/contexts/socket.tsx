import React, { PropsWithChildren, useContext } from "react";
import { Socket } from "socket.io-client";

const SocketContext = React.createContext<Socket | undefined>(undefined);

interface SocketProviderProps {
  socket: Socket;
}

export const SocketProvider = ({
  children,
  socket,
}: PropsWithChildren<SocketProviderProps>) => {
  return (
    <SocketContext.Provider value={socket}>{children}</SocketContext.Provider>
  );
};

export const useSocket = (): Socket => {
  const socket = useContext(SocketContext);

  if (socket === undefined) {
    throw new Error("useSocket() must be used within a SocketProvider");
  }

  return socket;
};
