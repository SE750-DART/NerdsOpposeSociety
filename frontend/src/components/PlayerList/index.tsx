import React, { useContext } from "react";
import createPersistedState from "use-persisted-state";
import Player from "./Player";
import styles from "./style.module.css";
import {
  Player as PlayerType,
  PlayersContext,
} from "../../ContextProviders/PlayersContextProvider";

const usePlayerIdState = createPersistedState("playerId");

interface PlayerListProps {
  gameState: "lobby" | "midround" | "endround" | "endgame";
}

const PlayerList = ({ gameState }: PlayerListProps) => {
  const { host, players } = useContext(PlayersContext);
  const [playerId] = usePlayerIdState("");

  const isHost = (player: PlayerType) => player.id === host;
  const isMe = (player: PlayerType) => player.id === playerId;

  let showScores: boolean = true;

  if (gameState === "lobby") {
    showScores = false;
  }

  return (
    <div className={styles.playerlist}>
      {players.map((player, index) => (
        <Player
          key={player.id}
          nickname={player.nickname}
          score={showScores ? player.score : undefined}
          highlight={isMe(player)}
          divider={
            index + 1 < players.length && // player not last
            !isMe(player) && // player not me
            !isMe(players[index + 1]) // next player not me
          }
          isHost={isHost(player)}
        />
      ))}
    </div>
  );
};

export default PlayerList;
