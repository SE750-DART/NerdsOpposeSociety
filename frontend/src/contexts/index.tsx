import React from "react";
import { PlayersContextProvider } from "./players";
import { PunchlinesContextProvider } from "./punchlines";
import { RoundContextProvider } from "./round";

const ContextProviders = ({ children }: { children: React.ReactNode }) => (
  <PlayersContextProvider>
    <PunchlinesContextProvider>
      <RoundContextProvider>{children}</RoundContextProvider>
    </PunchlinesContextProvider>
  </PlayersContextProvider>
);

export default ContextProviders;
