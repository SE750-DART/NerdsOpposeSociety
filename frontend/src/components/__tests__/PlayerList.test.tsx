import React from "react";
import renderer from "react-test-renderer";
import PlayerList from "../PlayerList";
import { PlayersContextProvider } from "../../contexts/players";

it("PlayerList snapshot testing", () => {
  const treeLobby = renderer
    .create(
      <PlayersContextProvider>
        <PlayerList gameState="lobby" />
      </PlayersContextProvider>
    )
    .toJSON();
  expect(treeLobby).toMatchSnapshot();

  const treeMidgame = renderer
    .create(
      <PlayersContextProvider>
        <PlayerList gameState="midround" />
      </PlayersContextProvider>
    )
    .toJSON();
  expect(treeMidgame).toMatchSnapshot();
});
