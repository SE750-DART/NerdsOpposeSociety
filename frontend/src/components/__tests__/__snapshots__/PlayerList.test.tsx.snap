// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList snapshot testing 1`] = `
<div
  className="playerlist"
/>
`;

exports[`PlayerList snapshot testing 2`] = `
<div
  className="playerlist"
/>
`;
